Создал openapi.yaml

cement-hools opened the pull request.

Commit:
Создал openapi.yaml

andrey-lepekhin commented on schemas/openapi.yaml:
```diff
@@ -0,0 +1,605 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/HTTPValidationError'
+  /api/v1/user/{user_id}/history:
```
Nit: Я вот думаю, может на это тоже пост запрос? Это вроде довольно персональные данные. Как я понял, с GET больше возможностей нахакать.

cement-hools replied:
как ты нахакаеешь, если это всего лишь id пользователя.
могу переделать, но мне кажется все через GET делают

andrey-lepekhin replied:
Давай оставим, я специально пометил как придирку (Nit), настаивать не буду

andrey-lepekhin commented on schemas/openapi.yaml:
```diff
@@ -0,0 +1,605 @@
+  title: Auth
+  description: Авторизация пользователей
+  version: 0.1.0
+paths:
```
Не хватает смены пароля из ТЗ

andrey-lepekhin commented on schemas/openapi.yaml:
```diff
@@ -0,0 +1,605 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/HTTPValidationError'
+  /api/v1/roles:
```
Не хватает ручек для 

- назначить такому-то юзеру такую-то роль
- проверить что у такого-то юзера есть такая-то роль

cement-hools replied:
проверить роль можно получив все роли пользователя

andrey-lepekhin replied:
Ты прав, тогда только первая строка актуальна

andrey-lepekhin commented on schemas/openapi.yaml:
```diff
@@ -0,0 +1,605 @@
+          type: string
+          format: date-time
+      description: Запрос на вход пользователя.
+    LoginUserResponse:
```
По идее мы же хотим еще токен возвращать на логине?

andrey-lepekhin commented on schemas/openapi.yaml:
```diff
@@ -0,0 +1,605 @@
+      tags:
+        - User
+      summary: Logout All
+      description: Выход пользователя из всех сеансов.
```
В ТЗ было вроде из всех *других* сеансов.
Плюс тут на входе нужен токен

cement-hools replied:
я подумал по user_id сможем завершить сеансы

andrey-lepekhin replied:
Да-да, у меня коммент именно про описание

andrey-lepekhin replied:
Хотя сорри, первая строка про описание, а вторая про необходимость авторизации.

Кстати, я так понял в апи это обычно делают через заголовки. Может так же и у нас сделаем, а не отдельным пармаетром?

cement-hools replied:
не когда так не делал. покажи пример

andrey-lepekhin replied:
https://stackoverflow.com/a/45471010/196171

https://swagger.io/docs/specification/authentication/bearer-authentication/

cement-hools replied:
Добавил правила с авторизацией по токену на ручки

Commit:
Добавил endpoint для смены пароля

andrey-lepekhin commented on schemas/openapi.yaml:
```diff
@@ -0,0 +1,605 @@
+          items:
+            $ref: '#/components/schemas/Role'
+      description: Ответ на получения всех ролей.
+    ValidationError:
```
Вижу неконсистентность, но может ты мне расскажешь почему так выбрал.

Когда получаешь 200, там есть на верхнем уровне success, error, data.
Я бы ожидал что при ошибке будут такие же узлы на верхнем уровне, просто error будет заполнен, и success false и data пустое. Это касается схем всех ошибок.

cement-hools replied:
по идеи так и должно быть. свагер делаю через фастапи. не знаю как там указать ошибочный респонс

andrey-lepekhin replied:
Вот здесь можно вручную переписывать и смотреть https://editor.swagger.io/

Условно, можешь скопировать то, что ты уже сделал туда и вручную поправить текст, чтобы было как надо.

andrey-lepekhin commented on schemas/openapi.yaml:
```diff
@@ -0,0 +1,605 @@
+  /api/v1/roles:
+    get:
+      tags:
+        - Role
```
Мне кажется для пользования этой апишкой тоже нужны токены авторизации. Ну или как-то явно громко пометить что это внутренняя апи и внешняя авторизация не нужна для нее, типа тока из локалки будем пускать) 
Хотя, что я такое говорю, из локалки несекьюрно пускать всех попало тоже)

cement-hools replied:
из локалки пускают только админов, которые добавляют фильмы. 
так я понял теорию.
а они в джанго админ ходят

andrey-lepekhin replied:
Опиши тогда пожалуйста где-то как пользователи будут получать авторизацию для испольования этих ручек

cement-hools replied:
описал

andrey-lepekhin replied:
Вижу ты в итоге их тоже под авторизацией токенами сделал, всё верно?

cement-hools replied:
да

andrey-lepekhin reviewed (changes requested):
Очень круто что так быстро сделал, есть комментарии

Commit:
Добавил endpoint для Добавить роль пользователю

cement-hools committed:
методы только для авторизованных пользователей

andrey-lepekhin commented on schemas/openapi.yaml:
```diff
@@ -0,0 +1,794 @@
+    ChangePasswordUserRequest:
+      title: ChangePasswordUserRequest
+      required:
+        - user_id
```
В других ручках для определения пользователя используется login. А почему тут именно user_id?

cement-hools replied:
логин нужен чтобы залогинтеться, пользователь же не знает свой id.
а когда меняем пароль пользователь уже залогинен и может передать id для того чтобы легче было его найти в БД

andrey-lepekhin replied:
И логин и юзер айди уникальны будут в таблице и на оба будут индексы скорее всего. 

То есть ускорения поиска БД не будет, а вот фронтедеру, который будет апишку реализовывать, нужно будет думать почему идентификатор пользователя в разных ручках разный.

cement-hools replied:
ну хорошо, сделаем логин. только фронтеру думаю все ровно, все ровно по документации будет писать

cement-hools replied:
сделал логин

andrey-lepekhin commented on schemas/openapi.yaml:
```diff
@@ -0,0 +1,794 @@
+    LogoutUser:
+      title: LogoutUser
+      required:
+        - refresh_token
```
Я пока не до конца понимаю как будет работать эта система, но на всякий случай подсвечу: тут точно рефреш токен на вход нужен? Я думал он нужен чтобы именно рефрешить. А для логаута нужен обычный токен.

Повторю, не знаю как это работает, это именно вопрос, не просьба что-то менять.

cement-hools replied:
я тоже не доконца понял, может и токен нужен.

andrey-lepekhin commented on schemas/openapi.yaml:
```diff
@@ -0,0 +1,794 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/HTTPValidationError'
+  /api/v1/user/{user_id}/roles/add:
```
Вот тут непонятно. Судя по расположению ручки, ты как будто предполагаешь что юзер сам будут себе добавлять роль? 

Я думал добавление ролей это какая-то внутренняя логика, не доступная внешнемю юзеру. То есть часть второй апишки roles.

Или я не понимаю чего-то. Можешь поясинть?

cement-hools replied:
Не сам будет добавлять. но по restfull думаю так правильнее. есть юзер, его роли. и добавление

andrey-lepekhin replied:
В ТЗ наделение ролями это часть второй апишки и я сколнен согласиться.
Если унести эту часть в апишку ролей, то вся папка users будет доступна извне и это будет понятное разделение.

Можно сделать рестфул через ресурс роли, например, roles/{role_id}/add_user

cement-hools replied:
хорошо

cement-hools replied:
сделал

andrey-lepekhin commented on schemas/openapi.yaml:
```diff
@@ -0,0 +1,794 @@
+            application/json:
+              schema:
+                $ref: '#/components/schemas/HTTPValidationError'
+  /api/v1/logout:
```
Тут кстати не нужна авторизация по токену?

cement-hools replied:
нужна!

cement-hools replied:
сделал

Commit:
Изменил usr_id на login
Перенес добавления пользователей в Role
Добавил ручку удалить пользователя из роли
авторизация для logout

Commit:
исправил описание к Role check user
